### PR TITLE
fix(pdk): use deep copies of Route, Service, and Consumer objects when log serialize

### PR DESCRIPTION
**Pull request by sabertobihwy**
### Summary

In the function of `kong.log.serialize`, the lifecycle of three table `ctx.route`, `ctx.service` and `ctx.authenticated_consumer` is across request. Modifying the sub-items in the current request of the table will affect the next request, resulting in unexpected behavior in Kong.

### Checklist

- [x] The Pull Request has tests
- [x] A changelog file has been added to `CHANGELOG/unreleased/kong` or adding `skip-changelog` label on PR if unnecessary. [README.md](https://github.com/Kong/kong/blob/master/CHANGELOG/README.md)
- [ ] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE

### Full changelog

* This commit modifies the code to utilize deep copies of the Route, Service, and Consumer objects in the log serializer. This ensures that the original objects are not modified when logging data.
* Add four test cases.

### Issue reference

Fix [FTI-5357](https://konghq.atlassian.net/browse/FTI-5357)

[FTI-5357]: https://konghq.atlassian.net/browse/FTI-5357?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: CHANGELOG/unreleased/kong/11566.yaml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+message: "use deep copies of Route, Service, and Consumer objects when log serialize"
```
Comment (chronolaw) [inline CHANGELOG/unreleased/kong/11566.yaml:1]:
```suggestion
message: "use deep copies of Route, Service, and Consumer objects when log serializing"
```

File: kong/pdk/log.lua [hunk 2]
```diff
@@ -832,9 +832,9 @@ do
         },
         tries = (ctx.balancer_data or {}).tries,
         authenticated_entity = build_authenticated_entity(ctx),
-        route = ctx.route,
-        service = ctx.service,
-        consumer = ctx.authenticated_consumer,
+        route = utils.cycle_aware_deep_copy(ctx.route),
```
Comment (chronolaw) [inline kong/pdk/log.lua:835]: Should we localize `utils.cycle_aware_deep_copy`?

File: spec/01-unit/10-log_serializer_spec.lua [hunk 3]
```diff
@@ -341,6 +355,20 @@ describe("kong.log.serialize", function()
 
         assert.is_nil(res.tries)
       end)
+
+      it("use the deep copies of the Route, Service, Consumer object avoid " .. 
+         "modify ctx.authenticated_consumer, ctx.route, ctx.service", function()
+        ngx.ctx.authenticated_consumer = {id = "someconsumer"}
```
Comment (chronolaw) [inline spec/01-unit/10-log_serializer_spec.lua:361]:
```suggestion
        ngx.ctx.authenticated_consumer = { id = "someconsumer" }
```

File: spec/02-integration/07-sdk/02-log_spec.lua [hunk 4]
```diff
@@ -121,3 +123,339 @@ describe("PDK: kong.log", function()
     f:close()
   end)
 end)
+
+for _, strategy in helpers.each_strategy() do
+  describe("PDK: kong.log.serialize [#" .. strategy .. "]", function()
+    describe("http subsystem 1", function()
```
Comment (chronolaw) [inline spec/02-integration/07-sdk/02-log_spec.lua:129]: Could we use more meaningful description for test suite?

Comment (sabertobihwy) [reply spec/02-integration/07-sdk/02-log_spec.lua]: update

File: spec/01-unit/10-log_serializer_spec.lua (hunk 3 above)
Comment (sabertobihwy) [reply spec/01-unit/10-log_serializer_spec.lua]: update

File: kong/pdk/log.lua (hunk 2 above)
Comment (sabertobihwy) [reply kong/pdk/log.lua]: update

File: CHANGELOG/unreleased/kong/11566.yaml (hunk 1 above)
Comment (sabertobihwy) [reply CHANGELOG/unreleased/kong/11566.yaml]: update

File: spec/02-integration/07-sdk/02-log_spec.lua [hunk 5]
```diff
@@ -121,3 +123,340 @@ describe("PDK: kong.log", function()
...
+          }
+
+          local body = assert.res_status(418, res)
+          local json = cjson.decode(body)
+          assert.equal(json["matched_route"]["name"], "route1")
+        end
+      end)
+    end)
+
+    describe("in strean subsystem# ctx.authenticated_consumer", function()
```
Comment (chronolaw) [inline spec/02-integration/07-sdk/02-log_spec.lua:384]:
```suggestion
    describe("in stream subsystem# ctx.authenticated_consumer", function()
```

Comment (sabertobihwy) [reply spec/02-integration/07-sdk/02-log_spec.lua]: fix

**Review by chronolaw — APPROVED** [review]
Great Job!

**Pull request merged**